File: src/restfw_admin/restfw_admin/validators.py
# -*- coding: utf-8 -*-
"""
:Authors: cykooz
:Date: 15.07.2020
"""

from dataclasses import dataclass, fields
from typing import ClassVar, List

from restfw.typing import JsonNumber, SimpleJsonValue

from .models import ValidatorModel


@dataclass()
class Validator:
    name: ClassVar[str]

    def to_model(self) -> ValidatorModel:
        args = tuple(getattr(self, f.name) for f in fields(self))
        return ValidatorModel(self.name, args)


class Required(Validator):
    name = 'required'


@dataclass()
class MinValue(Validator):
    name = 'minValue'
    min: JsonNumber


@dataclass()
class MaxValue(Validator):
    name = 'maxValue'
    max: JsonNumber


@dataclass()
class MinLength(Validator):
    name = 'minLength'
    min: int


@dataclass()
class MaxLength(Validator):
    name = 'maxLength'
    max: int


class Number(Validator):
    name = 'number'


class Email(Validator):
    name = 'email'


@dataclass()
class Regex(Validator):
    name = 'regex'
    pattern: str


@dataclass()
class Choices(Validator):
    name = 'choices'
    choices: List[SimpleJsonValue]

File: src/admin_ui/src/validators.ts
import {
    required,
    minLength,
    maxLength,
    minValue,
    maxValue,
    number,
    regex,
    email,
    choices
} from 'react-admin';
import {IField} from "./apiInfo";


function create_regex(pattern: string) {
    const re = new RegExp(pattern);
    return regex(re);
}


const VALIDATORS: Record<string, any> = {
    'required': required,
    'minValue': minValue,
    'maxValue': maxValue,
    'minLength': minLength,
    'maxLength': maxLength,
    'number': number,
    'email': email,
    'regex': create_regex,
    'choices': choices,
};


export function getFieldValidators(field: IField) {
    return field.validators.map((validator, index) => {
        if (validator.name in VALIDATORS) {
            return VALIDATORS[validator.name].apply(undefined, validator.args);
        }
        return null;
    }).filter((validator) => {
        return validator !== null
    });
}
